File: Solution Compare Nodes.js

function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @return {TreeNode}
 */
var inorderSuccessor = function (root, p) {
    this.p_successorInOrder = null;
    this.previous = null;
    return (p.right !== null) ? leftmostNodeFromCurrentRoot(p.right) : inorder(root, p);
};

/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @return {TreeNode}
 */
function inorder(root, p) {

    if (root === null) {
        return null;
    }
    if (this.p_successorInOrder !== null) {
        return this.p_successorInOrder;
    }

    inorder(root.left, p);
    if (this.previous === p) {
        this.p_successorInOrder = root;
    }
    this.previous = root;
    inorder(root.right, p);

    return this.p_successorInOrder;
}

/**
 * @param {TreeNode} node
 * @return {TreeNode}
 */
function leftmostNodeFromCurrentRoot(node) {
    while (node.left !== null) {
        node = node.left;
    }
    return node;
}

File: Solution Compare Nodes.cpp

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
    
public:
    TreeNode* p_successorInOrder{};
    TreeNode* previous{};

    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        return (p->right != nullptr) ? leftmostNodeFromCurrentRoot(p->right) : inorder(root, p);
    }

    TreeNode* inorder(TreeNode* root, TreeNode* p) {

        if (root == nullptr) {
            return nullptr;
        }
        if (p_successorInOrder != nullptr) {
            return p_successorInOrder;
        }

        inorder(root->left, p);
        if (previous == p) {
            p_successorInOrder = root;
        }
        previous = root;
        inorder(root->right, p);

        return p_successorInOrder;
    }

    TreeNode* leftmostNodeFromCurrentRoot(TreeNode* node) {
        while (node->left != nullptr) {
            node = node->left;
        }
        return node;
    }
};
